# 01. Import Libraries and Dataframes

In [2]:
import pandas as pd
import numpy as np
import os

In [4]:
path = r'C:\Users\bruna\Career Foundry\08-2023 Instacart Basket Analysis'

In [5]:
df_prods = pd.read_csv(os.path.join(path,'02 Data', 'Original Data','products.csv'), index_col=False)

In [5]:
df_ords = pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data','orders_wrangled.csv'), index_col=False)



# 02. Practice Fixing Mixed-Type Columns

In [6]:
# Create a dataframa

df_test = pd.DataFrame()

In [7]:
# Create a mixed type column

df_test['mix'] = ['a','b', 1, True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Check for mixed type columns

for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print (col)

mix


In [10]:
df_test['mix'] = df_test['mix'].astype('str')



# 03. Missing Values (df_prods)

In [11]:
# Finding out columns with missing values

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
# Creating a df to visualise columns with missing values only ('product name' column)

df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
# Checking initial number of rows before removing missing rows

df_prods.shape

(49693, 5)

In [15]:
# Creating a new dataframe without the missing rows

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [16]:
# Checking final number of rows after removing missing rows

df_prods_clean.shape

(49677, 5)



# 04. Duplicated Values (df_prods)

In [17]:
# Seeing duplicate values

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [18]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [19]:
# Checking initial number of rows before removing duplicates

df_prods_clean.shape

(49677, 5)

In [20]:
# Removing duplicates

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
# Checking final number of rows after removing duplicates

df_prods_clean_no_dups.shape

(49672, 5)



# 05. Exercise 4.5

## 05a. Continuation consistecy checks 'df_prods'

In [22]:
# Descriptive analysis

df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

Comments: max price is higher than expected. 99999 might be a number to symbolise missing info

In [23]:


# Subsetting to have a look at the products with price 99999

df_prods_price_99999 = df_prods_clean_no_dups[df_prods_clean_no_dups['prices'] == 99999]

In [24]:
df_prods_price_99999

product_id           product_name  aisle_id  department_id   prices
33666       33664  2 % Reduced Fat  Milk        84             16  99999.0

Comments: 99999 seems to be an error. Value can be replaced with median

In [25]:
# Finding median for 'prices' column

df_prods_clean_no_dups.median()

C:\Users\bruna\AppData\Local\Temp\ipykernel_33124\2543494676.py:3: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  df_prods_clean_no_dups.median()


product_id       24850.5
aisle_id            69.0
department_id       13.0
prices               7.1
dtype: float64

In [26]:
# Replacing '99999' with median
# TBC how to do it 

## 05b. Consistency checks 'df_ords'

In [27]:
# Descriptive analysis

df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  time_of_order  days_since_prior_order  
count        3.421083e+06   3.421083e+06            3.214874e+06  
mean         2.776219e+00   1.345202e+01            1.111484e+01  
std          2.046829e+00   4.226088e+00            9.206737e+00  
min          0.000000e+00   0.000000e+00            0.000000e+00  
25%          1.000000e+00   1.000000e+01            4.000000e+00  
50%          3.000000e+00   1.300000e+01            7.000000e+00  
75%          5.000000e+00   1.600000e+01            1.500000e+01  
max          6.000000e+00   2.300000e+01            3.000000e+01

Comments:
- a) Column 'Unnamed: 0' is the index. The data type can be changed to str
- b) Column 'order_id' and 'order_id' can also be changed to str
- c) Can 'order_number' be changed to str? TBC - 
- d) 'Orders_day_of_week' seems correct, with min 0, 50% 3, and max 6
- e) 'Time_of_order' seems correct, with min 0h, 50% 13h, and max 23h
- f) 'Days_since_prior_order' seems correct, with min 0 days, 50% 7 days, and max 30 days. This is compatible with mean 11 days

In [28]:
# Checking for mixed-type data

for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)
    

Comments: no mixed type-data found

In [29]:
# Checking for missing values

df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
time_of_order                  0
days_since_prior_order    206209
dtype: int64

Comments: only 'days_since_prior_order' has missing values. This would be because on the first time a new client places an order this wouldn't be filled in. This also matches the number of clients (max user_id is 206209)

In [30]:
# Confirming theory that 'days_since_prior_order' missing = 'order_number' 1
# Done by creating subset for 'days_since_prior_order' missing

df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [31]:
df_nan.head(30)

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0             0   2539329        1             1                   2   
11           11   2168274        2             1                   2   
26           26   1374495        3             1                   1   
39           39   3343014        4             1                   6   
45           45   2717275        5             1                   3   
50           50   2086598        6             1                   5   
54           54   2565571        7             1                   3   
75           75    600894        8             1                   6   
79           79    280530        9             1                   1   
83           83   1224907       10             1                   2   
89           89   2411567       11             1                   0   
97           97   1483446       12             1                   1   
103         103   2618231       13             1                   6   
116         116   2906263       14             1                   2   
130         130   2514324       15             1                   5   
153         153   1273676       16             1                   2   
160         160   1737705       17             1                   2   
201         201   2780889       18             1                   6   
208         208   1321974       19             1                   6   
218         218    947722       20             1                   1   
223         223   2625565       21             1                   1   
257         257   2300537       22             1                   3   
273         273   1792521       23             1                   5   
278         278    611214       24             1                   3   
297         297   2610576       25             1                   1   
301         301   3229714       26             1                   2   
314         314    361493       27             1                   3   
396         396   1138968       28             1                   0   
421         421   2064090       29             1                   5   
440         440    375303       30             1                   1   

     time_of_order  days_since_prior_order  
0                8                     NaN  
11              11                     NaN  
26              14                     NaN  
39              11                     NaN  
45              12                     NaN  
50              18                     NaN  
54               9                     NaN  
75               0                     NaN  
79              17                     NaN  
83              14                     NaN  
89              11                     NaN  
97              12                     NaN  
103             12                     NaN  
116             18                     NaN  
130             13                     NaN  
153              8                     NaN  
160             13                     NaN  
201             18                     NaN  
208             13                     NaN  
218             10                     NaN  
223             10                     NaN  
257              7                     NaN  
273             14                     NaN  
278             20                     NaN  
297             14                     NaN  
301             15                     NaN  
314              9                     NaN  
396             14                     NaN  
421             11                     NaN  
440             13                     NaN

In [32]:
# Address missing data


I've chosen not to act on the missing data, since the fact that this column has no data is significant. Replacing with another number would be misleading (ex: replacing with 0 would lead to conclusions that clients placed a new order on the same day) or impact future calculations; I also don't want to filter these values out, as they are not really missing. With that in mind, I've creates a new variable to act as a flag (ex: 'first_order_flag' with values True/False).

In [33]:
## - Adding a flag column 'first_order_flag'

df_ords['first_order_flag'] = df_ords['days_since_prior_order'].apply(lambda x: 'True' if pd.isna(x) else 'False')

In [34]:
df_ords.head(30)

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0            0   2539329        1             1                   2   
1            1   2398795        1             2                   3   
2            2    473747        1             3                   3   
3            3   2254736        1             4                   4   
4            4    431534        1             5                   4   
5            5   3367565        1             6                   2   
6            6    550135        1             7                   1   
7            7   3108588        1             8                   1   
8            8   2295261        1             9                   1   
9            9   2550362        1            10                   4   
10          10   1187899        1            11                   4   
11          11   2168274        2             1                   2   
12          12   1501582        2             2                   5   
13          13   1901567        2             3                   1   
14          14    738281        2             4                   2   
15          15   1673511        2             5                   3   
16          16   1199898        2             6                   2   
17          17   3194192        2             7                   2   
18          18    788338        2             8                   1   
19          19   1718559        2             9                   2   
20          20   1447487        2            10                   1   
21          21   1402090        2            11                   1   
22          22   3186735        2            12                   1   
23          23   3268552        2            13                   4   
24          24    839880        2            14                   3   
25          25   1492625        2            15                   1   
26          26   1374495        3             1                   1   
27          27    444309        3             2                   3   
28          28   3002854        3             3                   3   
29          29   2037211        3             4                   2   

    time_of_order  days_since_prior_order first_order_flag  
0               8                     NaN             True  
1               7                    15.0            False  
2              12                    21.0            False  
3               7                    29.0            False  
4              15                    28.0            False  
5               7                    19.0            False  
6               9                    20.0            False  
7              14                    14.0            False  
8              16                     0.0            False  
9               8                    30.0            False  
10              8                    14.0            False  
11             11                     NaN             True  
12             10                    10.0            False  
13             10                     3.0            False  
14             10                     8.0            False  
15             11                     8.0            False  
16              9                    13.0            False  
17             12                    14.0            False  
18             15                    27.0            False  
19              9                     8.0            False  
20             11                     6.0            False  
21             10                    30.0            False  
22              9                    28.0            False  
23             11                    30.0            False  
24             10                    13.0            False  
25             11                    30.0            False  
26             14                     NaN             True  
27             19                     9.0            False  
28             16                

In [35]:
# Checking for duplicate values

df_ords_dups = df_ords[df_ords.duplicated()]

In [36]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, time_of_order, days_since_prior_order, first_order_flag]
Index: []

No duplicates found

## 05c. Export cleaned dataframes

In [37]:
# Exporting df_prods

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'prods_checked.csv'))

In [38]:
# Exporting df_ords

df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'ords_checked.csv'))

In [1]:
df_prods_clean_no_dups.shape

NameError: name 'df_prods_clean_no_dups' is not defined